# Name: Chibueze Christian Nnaji
# Cohort: CML12

# Data Exploratory Analysis (EDA) and Classification Modeling Project

## 1. Load and read the data

We first load the data and read the first 5 rows of the data to get an insight on the type of data we are working with

Before we proceed, we need to import all the necessary packages we need

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

We import and load the data

In [18]:
filename = "phpB0xrNj.csv"
filepath = "C:/Users/chibueze.nnaji/Documents/CampusSquare/Projects/"
fullfilepath = filepath + filename
data = pd.read_csv(fullfilepath)

In [19]:
#Viewing the first 5 rows of the data
data.head()

f1      f2      f3      f4      f5      f6      f7      f8      f9  \
0 -0.4394 -0.0930  0.1718  0.4620  0.6226  0.4704  0.3578  0.0478 -0.1184   
1 -0.4348 -0.1198  0.2474  0.4036  0.5026  0.6328  0.4948  0.0338 -0.0520   
2 -0.2330  0.2124  0.5014  0.5222 -0.3422 -0.5840 -0.7168 -0.6342 -0.8614   
3 -0.3808 -0.0096  0.2602  0.2554 -0.4290 -0.6746 -0.6868 -0.6650 -0.8410   
4 -0.3412  0.0946  0.6082  0.6216 -0.1622 -0.3784 -0.4324 -0.4358 -0.4966   

      f10  ...    f609    f610    f611    f612    f613    f614    f615  \
0 -0.2310  ...  0.4102  0.2052  0.3846  0.3590  0.5898  0.3334  0.6410   
1 -0.1302  ...  0.0000  0.2954  0.2046  0.4772  0.0454  0.2046  0.4318   
2 -0.8318  ... -0.1112 -0.0476 -0.1746  0.0318 -0.0476  0.1112  0.2540   
3 -0.9614  ... -0.0504 -0.0360 -0.1224  0.1366  0.2950  0.0792 -0.0072   
4 -0.5406  ...  0.1562  0.3124  0.2500 -0.0938  0.1562  0.3124  0.3124   

     f616    f617  class  
0  0.5898 -0.4872    '1'  
1  0.4546 -0.0910    '1'  
2  0.1588 -0.4762    '2'  
3  0.0936 -0.1510    '2'  
4  0.2188 -0.2500    '3'  

[5 rows x 618 columns]

In [20]:
#getting insight in our data
data.describe()

f1           f2           f3           f4           f5  \
count  7797.000000  7797.000000  7797.000000  7797.000000  7797.000000   
mean     -0.385396     0.141103     0.347384     0.450659     0.324283   
std       0.236721     0.320090     0.326658     0.308301     0.461491   
min      -1.000000    -0.892600    -0.975200    -0.968000    -0.996600   
25%      -0.546800    -0.099400     0.105400     0.247800    -0.024800   
50%      -0.414000     0.106400     0.333400     0.453200     0.415800   
75%      -0.241000     0.354600     0.583400     0.668600     0.692400   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                f6           f7           f8           f9          f10  ...  \
count  7797.000000  7797.000000  7797.000000  7797.000000  7797.000000  ...   
mean      0.163589     0.023852    -0.081577    -0.149191    -0.191763  ...   
std       0.569970     0.559704     0.535142     0.528714     0.515343  ...   
min      -1.000000    -1.000000    -1.000000    -1.000000    -1.000000  ...   
25%      -0.359600    -0.461600    -0.526800    -0.577400    -0.601200  ...   
50%       0.204200    -0.010800    -0.154400    -0.233200    -0.261000  ...   
75%       0.686200     0.492400     0.304400     0.190200     0.120000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

              f608         f609         f610         f611         f612  \
count  7797.000000  7797.000000  7797.000000  7797.000000  7797.000000   
mean      0.370768     0.341730     0.311678     0.318904     0.334451   
std       0.335582     0.330835     0.328813     0.333747     0.334954   
min      -0.609400    -0.803400    -0.775200    -0.830200    -0.711800   
25%       0.130400     0.098000     0.071400     0.073600     0.088600   
50%       0.366200     0.337400     0.294200     0.295000     0.315800   
75%       0.622600     0.588800     0.548800     0.565800     0.581600   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

              f613         f614         f615         f616         f617  
count  7797.000000  7797.000000  7797.000000  7797.000000  7797.000000  
mean      0.299238     0.224122     0.186668     0.081370    -0.281416  
std       0.334282     0.336135     0.337780     0.332859     0.354418  
min      -0.770800    -0.810600    -0.780200    -0.802800    -0.962600  
25%       0.049200    -0.017800    -0.057600    -0.154600    -0.546600  
50%       0.280000     0.192400     0.157800     0.045200    -0.329000  
75%       0.541200     0.461600     0.410800     0.289800    -0.065600  
max       1.000000     1.000000     1.000000     1.000000     0.889800  

[8 rows x 617 columns]

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7797 entries, 0 to 7796
Columns: 618 entries, f1 to class
dtypes: float64(613), int64(4), object(1)
memory usage: 36.8+ MB


In [12]:
data.isna().sum()

f1       0
f2       0
f3       0
f4       0
f5       0
        ..
f614     0
f615     0
f616     0
f617     0
class    0
Length: 618, dtype: int64

## 2. Using the pop function to separate the data 

In [21]:
data.columns

Index(['f1', 'f2', 'f3', 'f4', 'f5', 'f6', 'f7', 'f8', 'f9', 'f10',
       ...
       'f609', 'f610', 'f611', 'f612', 'f613', 'f614', 'f615', 'f616', 'f617',
       'class'],
      dtype='object', length=618)

In [22]:
#Spliting the data using drop and pop functions
X = data.drop(columns='class')
Y = data.pop('class')

In [23]:
X.head()

f1      f2      f3      f4      f5      f6      f7      f8      f9  \
0 -0.4394 -0.0930  0.1718  0.4620  0.6226  0.4704  0.3578  0.0478 -0.1184   
1 -0.4348 -0.1198  0.2474  0.4036  0.5026  0.6328  0.4948  0.0338 -0.0520   
2 -0.2330  0.2124  0.5014  0.5222 -0.3422 -0.5840 -0.7168 -0.6342 -0.8614   
3 -0.3808 -0.0096  0.2602  0.2554 -0.4290 -0.6746 -0.6868 -0.6650 -0.8410   
4 -0.3412  0.0946  0.6082  0.6216 -0.1622 -0.3784 -0.4324 -0.4358 -0.4966   

      f10  ...    f608    f609    f610    f611    f612    f613    f614  \
0 -0.2310  ...  0.3334  0.4102  0.2052  0.3846  0.3590  0.5898  0.3334   
1 -0.1302  ...  0.2272  0.0000  0.2954  0.2046  0.4772  0.0454  0.2046   
2 -0.8318  ...  0.0952 -0.1112 -0.0476 -0.1746  0.0318 -0.0476  0.1112   
3 -0.9614  ...  0.0648 -0.0504 -0.0360 -0.1224  0.1366  0.2950  0.0792   
4 -0.5406  ...  0.2812  0.1562  0.3124  0.2500 -0.0938  0.1562  0.3124   

     f615    f616    f617  
0  0.6410  0.5898 -0.4872  
1  0.4318  0.4546 -0.0910  
2  0.2540  0.1588 -0.4762  
3 -0.0072  0.0936 -0.1510  
4  0.3124  0.2188 -0.2500  

[5 rows x 617 columns]

In [24]:
Y

0        '1'
1        '1'
2        '2'
3        '2'
4        '3'
        ... 
7792    '24'
7793    '25'
7794    '25'
7795    '26'
7796    '26'
Name: class, Length: 7797, dtype: object

## 3. Spliting the data into train and test

In [25]:
from sklearn.model_selection import train_test_split

In [33]:
#Spliting the data into train and test data
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.3,random_state=123,stratify=Y)

In [31]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5457 entries, 5222 to 5351
Columns: 617 entries, f1 to f617
dtypes: float64(613), int64(4)
memory usage: 25.7 MB


In [32]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2340 entries, 1785 to 7511
Columns: 617 entries, f1 to f617
dtypes: float64(613), int64(4)
memory usage: 11.0 MB


## 4. Training the model with RandomForestClassifier

In [34]:
#importing the RandomForestClassifier algorithm
from sklearn.ensemble import RandomForestClassifier

In [35]:
model = RandomForestClassifier()

In [36]:
#Training our model
model.fit(X_train,Y_train)

RandomForestClassifier()

In [39]:
#Testing our model with the train data
score = model.score(X_train,Y_train)
score

1.0

In [40]:
#Testing our model with data it hasnt seen before
score1 = model.score(X_test,Y_test)
score1

0.9358974358974359

## 5. Using our trained model to predict

In [47]:
#Predicting using the train data
pred1 = model.predict(X_train)
pred1

array(["'23'", "'1'", "'5'", ..., "'4'", "'25'", "'23'"], dtype=object)

In [48]:
#predicing using the test data
pred2 = model.predict(X_test)
pred2

array(["'9'", "'22'", "'13'", ..., "'21'", "'7'", "'14'"], dtype=object)

## 6. Checking the accuracy of our trained model

In [43]:
#Importing the accuracy score function to check the accuracy of our trained model
from sklearn.metrics import accuracy_score

In [50]:
#Checking the accuracy of model on the train data
accuracy1 = accuracy_score(Y_train,pred1)
accuracy1

1.0

In [52]:
#Checking the accuracy of our model on the test data
accuracy2 = accuracy_score(Y_test,pred2)
accuracy2

0.9358974358974359

## 7. Hyperparameter tuning

In [55]:
#Using n_estimators = 20,random_state = 123 and verbose =2
model1 = RandomForestClassifier(n_estimators=20,random_state=123,verbose=2)

In [56]:
model1.fit(X_train,Y_train)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s


building tree 1 of 20
building tree 2 of 20
building tree 3 of 20
building tree 4 of 20
building tree 5 of 20
building tree 6 of 20
building tree 7 of 20
building tree 8 of 20
building tree 9 of 20
building tree 10 of 20
building tree 11 of 20
building tree 12 of 20
building tree 13 of 20
building tree 14 of 20
building tree 15 of 20
building tree 16 of 20
building tree 17 of 20
building tree 18 of 20
building tree 19 of 20
building tree 20 of 20


[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    1.8s finished


RandomForestClassifier(n_estimators=20, random_state=123, verbose=2)

In [96]:
model1.score(X_test,Y_test)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:    0.0s finished


0.9213675213675213

In [67]:
#Using n_estimators=20,max_depth=5,min_samples_leaf=10,max_features=0.5,random_state=123
model2 = RandomForestClassifier(n_estimators=20,max_depth=5,min_samples_leaf=10,max_features=0.5,random_state=123)

In [68]:
model2.fit(X_train,Y_train)

RandomForestClassifier(max_depth=5, max_features=0.5, min_samples_leaf=10,
                       n_estimators=20, random_state=123)

In [69]:
model2.score(X_test,Y_test)

0.7495726495726496

In [86]:
predict2 = model2.predict(X_test)
predict2

array(["'9'", "'4'", "'12'", ..., "'21'", "'7'", "'14'"], dtype=object)

In [87]:
accuracy_score2 = accuracy_score(Y_test,predict2)
accuracy_score2

0.7495726495726496

In [70]:
#Using n_estimators=20,max_depth=10,min_samples_leaf=50,max_features=0.3,random_state=123
model3 = RandomForestClassifier(n_estimators=20,max_depth=10,min_samples_leaf=50,max_features=0.3,random_state=123)

In [71]:
model3.fit(X_train,Y_train)

RandomForestClassifier(max_depth=10, max_features=0.3, min_samples_leaf=50,
                       n_estimators=20, random_state=123)

In [72]:
model3.score(X_test,Y_test)

0.8670940170940171

In [88]:
predict3 = model3.predict(X_test)
predict3

array(["'9'", "'22'", "'13'", ..., "'21'", "'7'", "'14'"], dtype=object)

In [89]:
accuracy_score3 = accuracy_score(Y_test,predict3)
accuracy_score3

0.8670940170940171

In [73]:
#Using n_estimators=50,max_depth=10,min_samples_leaf=50,max_features=0.5,random_state=123
model4 = RandomForestClassifier(n_estimators=50,max_depth=10,min_samples_leaf=50,max_features=0.5,random_state=123)

In [74]:
model4.fit(X_train,Y_train)

RandomForestClassifier(max_depth=10, max_features=0.5, min_samples_leaf=50,
                       n_estimators=50, random_state=123)

In [75]:
model4.score(X_test,Y_test)

0.8658119658119658

In [90]:
predict4 = model4.predict(X_test)
predict4

array(["'9'", "'22'", "'13'", ..., "'21'", "'7'", "'14'"], dtype=object)

In [91]:
accuracy_score4 = accuracy_score(Y_test,predict4)
accuracy_score4

0.8658119658119658

In [78]:
# Using n_estimators=150,max_depth=20,min_samples_leaf=100,max_features=1,random_state=123
model5 = RandomForestClassifier(n_estimators=150,max_depth=20,min_samples_leaf=100,max_features=1,random_state=123)

In [80]:
model5.fit(X_train,Y_train)

RandomForestClassifier(max_depth=20, max_features=1, min_samples_leaf=100,
                       n_estimators=150, random_state=123)

In [81]:
model5.score(X_test,Y_test)

0.8444444444444444

In [92]:
predict5 = model5.predict(X_test)
predict5

array(["'9'", "'22'", "'12'", ..., "'17'", "'7'", "'14'"], dtype=object)

In [93]:
accuracy_score5 = accuracy_score(Y_test,predict5)
accuracy_score5

0.8444444444444444

In [82]:
#Using the default configuration
model6 = RandomForestClassifier()

In [83]:
model6.fit(X_train,Y_train)

RandomForestClassifier()

In [84]:
model6.score(X_test,Y_test)

0.9376068376068376

In [94]:
predict6 = model6.predict(X_test)
predict6

array(["'9'", "'22'", "'13'", ..., "'21'", "'7'", "'14'"], dtype=object)

In [95]:
accuracy_score6 = accuracy_score(Y_test,predict6)
accuracy_score6

0.9376068376068376

## 8.Selecting the best hyperparameter value

1. First model(model1) I created with hyperparamters of n_estimators = 20,random_state = 123 and verbose = 2 gave an accuracy_score of 92.14%

2. Second model(model2) I created with hyperparameters of n_estimators=20,max_depth=5,min_samples_leaf=10,max_features=0.5,random_state=123 gave an accuracy_score of 74.96%

3. Third model(model3) I created with hyperparameters of n_estimators=20,max_depth=10,min_samples_leaf=50,max_features=0.3,random_state=123 gave an accuracy_score of 86.71%

4. Fourth model(model4) I created with hyperparameters of n_estimators=50,max_depth=10,min_samples_leaf=50,max_features=0.5,random_state=123 gave an accuracy_score of 86.58%

5. Fifth model(model5) I created with hyperparameters of n_estimators=150,max_depth=20,min_samples_leaf=100,max_features=1,random_state=123 gave an accuracy_score of 84.44%

6. Sixth model(model6) I created with default parameters in the RandomForestClassifier hyperparameters with default values of  n_estimators=100,
    criterion='gini',
    max_depth=None,
    min_samples_split=2,
    min_samples_leaf=1,
    min_weight_fraction_leaf=0.0,
    max_features='auto',
    max_leaf_nodes=None,
    min_impurity_decrease=0.0 
    and I got an accuracy_score of 93.76%

The best model was the sixth model(model6) I created with the default hyperparameters which has an accuracy_score of 93.76% followed by First model(model1) I created with hyperparameters of n_estimators = 20,random_state = 123 and verbose = 2 with an accuracy_score of 92.14% and Second model(model2) I created with hyperparameters of n_estimators=20,max_depth=5,min_samples_leaf=10,max_features=0.5,random_state=123 gave the least accuracy_score of 74.96%

## 9. EDA Data and Model Analysis

The EDA data I imported was already cleaned (no missing values) and it has already be rescaled(all the features have a minimum and maximum value of 0 and 1 respectively). So the data is cleaned, rescaled and is near normal distribution

The model was initially tested with the data it was trained with and we got an accuracy_score of 100% and when tested with data it has not seen before, the accuracy_score dropped to 93% which shows our model was initially overfitting.


The hyperparameter tuning has alot of significant in our model's performance. With the right hyperparamters, we can get the best performance from our model 
